linkerd/opencensus/src/lib.rs

```diff
@@ -1,59 +1,77 @@
-#![deny(warnings, rust_2018_idioms)]
-
-use futures::{task, try_ready, Async, Future, Poll, Stream};
+// #![deny(warnings, rust_2018_idioms)]
+use futures::{ready, Stream, TryStream, TryStreamExt};
+use http_body::Body as HttpBody;
 use linkerd2_error::Error;
+use linkerd2_stack::NewService;
 use metrics::Registry;
 pub use opencensus_proto as proto;
 use opencensus_proto::agent::common::v1::Node;
 use opencensus_proto::agent::trace::v1::{
-    client::TraceService, ExportTraceServiceRequest, ExportTraceServiceResponse,
+    trace_service_client::TraceServiceClient, ExportTraceServiceRequest, ExportTraceServiceResponse,
 };
 use opencensus_proto::trace::v1::Span;
+use pin_project::{pin_project, project};
 use std::convert::TryInto;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
 use tokio::sync::mpsc;
-use tower_grpc::{
-    self as grpc, client::streaming::ResponseFuture, generic::client::GrpcService, BoxBody,
+use tonic::{
+    self as grpc,
+    body::{Body as GrpcBody, BoxBody},
+    client::GrpcService,
+    Streaming,
 };
 use tracing::trace;
 
 pub mod metrics;
 
 /// SpanExporter sends a Stream of spans to the given TraceService gRPC service.
-pub struct SpanExporter<T, S>
-where
-    T: GrpcService<BoxBody>,
-{
+#[pin_project]
+pub struct SpanExporter<T, S> {
     client: T,
     node: Node,
-    state: State<T>,
+    state: State,
+    #[pin]
     spans: S,
     max_batch_size: usize,
     metrics: Registry,
 }
 
-enum State<T: GrpcService<BoxBody>> {
+enum State {
     Idle,
     Sending {
         sender: mpsc::Sender<ExportTraceServiceRequest>,
         // Node data should only be sent on the first message of a streaming
         // request.
         node: Option<Node>,
         // We hold the response future, but never poll it.
-        _rsp: ResponseFuture<ExportTraceServiceResponse, T::Future>,
+        _rsp: Pin<
```

Comment:
Lucio Franco mentioned to me that not polling this future _may_ result in the h2 connection window filling up, so I may look into changing this code so that it is polled (even though we don't care about the result).